### ISSUE TITLE
Post-training Llama-8B Models using different hyperparam configs on Tulu Datasets

### ISSUE THREAD

**Issue by HannanJaved** [OpenEuroLLM/Taskboard#101]
Llama3.1-8B-Base models were **LoRA** fine-tuned on the Tulu-3-SFT dataset mix and code to check how different learning rates, warmup ratios, and LoRA ranks affect performance of fine-tuned models. The models were finetuned on **4 H100** GPUs for **one** epoch.

# SFT Experimental Setting:
Configurations mixed using:
**Base Model:** Llama3.1-8B
**Dataset:** Tulu-3-SFT-mix
**LoRA Ranks** (LoRA alpha is set as Rank/4) = {64, 256, 1024}
**Learning Rate (LR)** = {1e-6, 1e-5, 5e-5}
**Warmup Ratio (WR)** = {0.001, 0.005, 0.01}

And one with the Default hyperparameter setting from the Tulu3 codebase

# Results
The finetuned models were then tested using:
**Benchmark:** Arena-Hard-v2.0
**No. of Instructions:** 750
**Baselines:** {Llama3.1-8B, Llama3.1-8B-Instruct, Tulu-Llama3.1-8B}
**Judge Model:** neuralmagic/Meta-Llama-3.1-70B-Instruct-FP8

_It is important to keep in mind that the Llama3.1-8B-Instruct and Tulu-Llama3.1-8B had a more complex post-training pipeline (e.g. RLVR), whereas our models have only been fine-tuned using SFT on the Tulu3-SFT-Dataset for 1 epoch using the Tulu3 codebase_ 

## Compared to Llama3.1-8B
<img width="1798" height="498" alt="Image" src="https://github.com/user-attachments/assets/c1619cf5-36f4-4f05-9474-d933499d3cd5" />

## Compared to Llama3.1-8B-Instruct
<img width="1798" height="498" alt="Image" src="https://github.com/user-attachments/assets/113be878-ce02-465b-b3e4-b2db7e34f0da" />

## Compared to Tulu-Llama3.1-8B
<img width="1798" height="498" alt="Image" src="https://github.com/user-attachments/assets/fc8eccb9-9c77-4e0e-93af-bf9768141ac2" />

# Next Steps

- [x] plot spearmanr across training iterations - to find the optimal rung sizes
- [x] Plot validation loss curves for the finetuned models
- [ ] Train models with same Rank, LR, and WarmupRatio but varying levels of LoRA alpha - is there an optimum LoRA alpha to rank relationship?
- [x] Perform DPO on good, bad, and average performing SFT models to see if better performing SFT models also have better final model performance after RL?
- [x] Perform SFT on some of the best checkpoints for one more epoch
- [x] Compare performance of checkpoints from previous point with the Llama-Tulu-8B-SFT model (without RL)
- [ ] change the dataset mixtures - possible combining with MF-BO

**Comment by HannanJaved** [comment]
# Plots of Spearman Correlation 

## Between Arena-Hard Score and Training Iterations

These figures show, for three Ranks (64, 256 and 1024), how well the ordering by win rates of models early in training predicts their final ordering by win rates. For each training iteration shown on the x‑axis, we collect every checkpoint that has both a score at that iteration and a final score (score=win_rate). We then compute a single number, the Spearman correlation, that measures how closely the ranking by score at that iteration matches the ranking by final score (1 = perfect match, 0 = no relationship). 

Each curve is the Spearman correlation over time for one rank; higher values mean earlier scores are better predictors of final performance. We do this to identify potentially "good" levels where early evaluations predict final outcomes reliably for later implementation of multi-fidelity optimization methods.
Basically measures how well performance at an early/mid-training checkpoint predicts the final performance of the model.

**Technicalities:**
For a given rank (64, 256, or 1024) and a given training step (e.g., 6000, 12000, …), the script collects every model configuration that has:
a score at that intermediate step (at a given training iteration or checkpoint), and
a corresponding final score (the final checkpoint).

It then produces two orderings (rankings):
ordering A: models sorted by their intermediate-score at that step (highest win rate first),
ordering B: the same models sorted by their final win rate (highest final win first).
The Spearman correlation measures how similar ordering A and ordering B are — high Spearman means the relative order of models early in training matches their relative order at the end.

### Against Base Llama3.1-8b

<img width="989" height="590" alt="Image" src="https://github.com/user-attachments/assets/cb5e40bb-ba06-4687-8e43-127692f387e7" />

### Against Llama3.1-8b-Instruct

<img width="989" height="590" alt="Image" src="https://github.com/user-attachments/assets/2851c1ab-221e-439b-bda6-80f46567dd18" />

### Against Llama3.1-8b finetuned by team at Tulu

<img width="989" height="590" alt="Image" src="https://github.com/user-attachments/assets/246b2e06-9d3b-4c3f-890f-967021500a7f" />

**Comment by geoalgo** [comment]
Thanks for sharing!
Could you precise what is your setup here? (It is not clear what are the two distributions being compared)

**Comment by HannanJaved** [comment]
Todos: 
- [ ] Update previous message on spearmanr - add details
- [ ] Plot validation loss curves for the finetuned models
- [ ] Do DPO on good, bad, and average performing SFT models to see if better performing SFT models also have better final model performance after RL

**Comment by HannanJaved** [comment]
# Validation Loss Curves per Rank

The models are validated on a random 5% split of the Tulu3-SFT-mix (same 5% used for validating all the checkpoints - was done post-hoc and not during training due to issues with the Tulu codebase)

Contrary to the Arena-hard scores, which peak at around step 36000 and then either degrade or plateau, the validation loss keeps dropping.

We can also see from these curves that the Warmup Ratio does not play a huge role in the validation performance of the models as they all seem to converge after a certain point in training.  Lower Warmup Ratios seem to start off at a lower validation loss though (see loss at training iteration 6000)

Learning Rate, on the other hand, has a bigger impact. Lowest LR of 1e-6 consistently shows the worst performance (both in terms of validation loss and AH scores)


## Rank 64

<img width="855" height="547" alt="Image" src="https://github.com/user-attachments/assets/3af0c79c-095f-49f1-9db0-40494dc496c9" />

## Rank 256

<img width="855" height="547" alt="Image" src="https://github.com/user-attachments/assets/7488d4a2-7522-4de7-a79d-f5d5123f780e" />

## Rank 1024

<img width="846" height="547" alt="Image" src="https://github.com/user-attachments/assets/24218bf1-b646-4d3e-bd94-36cf10f36989" />

**Comment by geoalgo** [comment]
Great results, thanks for sharing!

Looking at the loss it makes sense they stopped at two epochs.
Regarding the hyperparameters we are varying, the warmup ratio does not have a strong effect.

One hyperparameter we could replace is the coefficient for the KL regularization with the initial predictions (This one is likely quite important).

**Comment by HannanJaved** [comment]
# Correlation between the Validation Loss and the Arena-Hard Scores (vs Base Llama Model)

The plot suggests that validation loss on the Tulu-3 SFT mixture strongly predicts Arena-Hard performance across the model checkpoints. The idea was to check if validation loss on the Tulu-3-SFT-dataset can be a reliable proxy for downstream performance (Arena-hard v2 win rates)

This correlation holds consistently across: (1) different LoRA ranks (64, 256, 1024) shown in top-right, (2) different learning rates (1e-5, 5e-5, 1e-6) shown in bottom-left, and (3) different training stages shown in bottom-right. Training stages are defined as: Early (≤18,000 steps), Mid (18,001-36,000 steps), Late (36,001-48,000 steps), and Final (55,000 steps). 

Pearson and Spearman correlations were computed independently for each subgroup, with correlation coefficients displayed in the top-right corner of each panel. 

All subgroups maintain negative correlations, showing that as validation loss decreases, the Arena-Hard performance increases, regardless of model configuration or training progress. This suggests that we could possibly use validation loss as an early indicator of model quality without Arena-Hard evaluations.

<img width="4770" height="3597" alt="Image" src="https://github.com/user-attachments/assets/899971da-7840-47e6-848f-2f49b7962285" />

**Comment by geoalgo** [comment]
Very nice result and analysis! What is stricking is that for lower validation losses AH performance stalls.
This is very interesting, I wonder what will happen with DPO (hopefully this happens less). 
We can do the same analysis once we have the checkpoint of Ali!

**Comment by JeniaJitsev** [comment]
We can also consider cheaper experiments on smaller scales, eg see this using Tulu3 : https://arxiv.org/abs/2509.24945

If I understand this right, so far all experiments were LoRA based. We should also test conventional post-training without LoRA.

**Comment by HannanJaved** [comment]
> We can also consider cheaper experiments on smaller scales, eg see this using Tulu3 : https://arxiv.org/abs/2509.24945
>
> If I understand this right, so far all experiments were LoRA based. We should also test conventional post-training without LoRA.

Yes, these results are all on LoRA, one of David's colleagues (Ali) is looking into full finetuning as well.
Will check out the paper you shared too, thanks!
